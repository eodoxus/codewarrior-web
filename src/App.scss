@import "./lib/lib.scss";

.app {
  @include fill-container;

  .content {
    position: absolute;
    top: $dimension-header--height;
    right: 0;
    bottom: $dimension-footer--height;
    left: 0;

    .error {
      position: relative;
      width: 100%;
      padding: $gutter-sm;
      border: 1px solid $color-error--border;
      background: $color-error--background;
      color: $color-font-inverse;
      margin: $gutter-sm;
    }
  }

  .game {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -240px;
    margin-left: -320px;

    .border {
      position: absolute;
      top: -14px;
      left: -14px;
      background-size: 100%;
      width: 668px;
      height: 500px;
      z-index: 2;
    }
  }
}
