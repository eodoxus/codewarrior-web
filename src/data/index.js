import AppModel from "./AppModel";
import DataCollection from "./DataCollection";
import DataModel from "./DataModel";

export { AppModel, DataCollection, DataModel };
