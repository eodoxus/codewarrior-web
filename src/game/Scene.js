import React, { Component } from "react";
import styles from "./Scene.scss";

const TILE_SIZE = 32;

export default class Scene extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "Base Scene",
      isLoading: false
    };
  }

  onClick(x, y) {
    console.log("scene click", x, y);
  }

  render() {
    return <span>{this.state.name}</span>;
  }

  renderGrid() {}

  renderRow(row, y) {
    const cells = row.map((col, x) => (
      <div className={styles.cell} key={x + "-" + y}>
        {x}, {y}
      </div>
    ));
    return (
      <div className={styles.row} key={y}>
        {cells}
      </div>
    );
  }
}

Scene.defaultProps = {
  top: 100,
  left: 0,
  height: 400,
  width: 400
};
