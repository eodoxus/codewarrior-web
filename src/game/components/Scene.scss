$border: 1px solid black;

.grid {
  border: $border;
  display: flex;
  flex-direction: column;

  .row {
    border-bottom: $border;
    display: flex;
    flex-direction: row;
    width: 100%;

    .cell {
      border-right: $border;
    }
  }
}
