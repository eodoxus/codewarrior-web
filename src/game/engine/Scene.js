import Audio from "./Audio";
import entities from "../entities";
import Event from "../../lib/Event";
import Graphics from "./Graphics";
import Hero from "../entities/hero/Hero";
import Tile from "./map/Tile";
import TiledMap from "./map/TiledMap";

export default class Scene {
  clickedTile;
  name;
  hero;
  map;
  size;
  entities;

  constructor(hero) {
    Audio.stop();
    this.hero = hero;
    this.map = new TiledMap(this.getName());
    this.hero.setMap(this.map);
    this.entities = [hero];
  }

  addEntity(entity) {
    this.entities.push(entity);
  }

  getBackgroundMusic() {
    // Override this
  }

  getName() {
    // Override this
    return this.name;
  }

  getMap() {
    return this.map;
  }

  setMap(map) {
    this.map = map;
    this.entities.forEach(entity => {
      entity.setMap(this.map);
    });
  }

  getEntities() {
    return this.entities;
  }

  getSize(s) {
    return this.size;
  }

  setSize(s) {
    this.size = s;
  }

  detectCollisions(entity) {
    const tile = this.map.getTileAt(entity.getOrigin());
    if (tile && tile.isDoorway() && entity.isHero()) {
      return Event.fire(Event.DOORWAY, tile);
    }

    this.entities.forEach(nextEntity => {
      if (entity.getId() === nextEntity.getId()) {
        return;
      }
      if (
        (entity.getVelocity().magnitude() ||
          nextEntity.getVelocity().magnitude()) &&
        entity.intersects(nextEntity)
      ) {
        entity.handleCollision(nextEntity);
      }
    });
  }

  async loadAssets() {
    if (this.map) {
      await this.map.loadAssets();
      this.map.getEntities().forEach(tile => {
        const entityName = tile.getProperty(Tile.PROPERTIES.ENTITY);
        if (!entities[entityName]) {
          throw new Error(`Entity ${entityName} does not exist`);
        }
        const entity = new entities[entityName](
          tile.getProperty(Tile.PROPERTIES.NAME),
          tile.getPosition()
        );
        entity.setProperties(tile.getProperties());
        this.entities.push(entity);
      });
    }
    const promises = [];
    this.getEntities().forEach(entity => promises.push(entity.loadAssets()));
    const music = this.getBackgroundMusic();
    if (music) {
      Audio.play(music);
    }
    await Promise.all(promises);
  }

  onClick(position) {
    this.clickedPosition = position;
    this.clickedTile = this.map.getTileAt(position);
    if (this.clickedTile && this.clickedTile.isWalkable()) {
      this.hero.walkTo(this.clickedTile);
    }
  }

  render() {
    if (this.map) {
      this.map.render();

      if (Graphics.debug) {
        if (this.clickedPosition) {
          Graphics.drawPoint(this.clickedPosition);
        }
        if (this.clickedTile) {
          Graphics.colorize(this.clickedTile.getRect(), "black");
        }
      }
    }
    this.entities.forEach(entity => entity.render());
  }

  renderDebug() {
    return this.entities.map(entity => entity.renderDebug());
  }

  shouldShowBorder() {
    return this.showBorder;
  }

  spawnHero(position, direction) {
    if (!position) {
      position = this.map.getHeroSpawnPoint();
    }
    this.hero.setPosition(position);
    if (direction) {
      this.hero.getSprite().pickAnimation(Hero.STATES.WALKING, direction);
    }
  }

  update(dt) {
    this.map.trackEntities(this.entities);
    this.entities.forEach(entity => {
      entity.update(dt);
      this.detectCollisions(entity);
    });
  }

  unload() {
    this.hero.stop();
  }
}
