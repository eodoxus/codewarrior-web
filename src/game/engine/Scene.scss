@import "../../lib/lib.scss";

.scene {
  background: transparent;
}

.sprite {
  background: transparent;
}
