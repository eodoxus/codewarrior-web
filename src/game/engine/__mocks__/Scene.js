import Texture from "../Texture";
import Vector from "../Vector";
import Size from "../Size";
import TiledMap from "../map/TiledMap";

import tmxConfig from "../map/__mocks__/map.json";
const map = new TiledMap("testMap");
map.loadTMXConfig(tmxConfig);

export default class Scene {
  constructor(name) {
    this.name = name;
  }
  getEntities() {
    return [this.hero];
  }
  init = jest.fn();
  addEntity = entity => {
    if (entity.isHero()) {
      this.hero = entity;
    }
  };
  getApi = jest.fn();
  getBackgroundMusic = () => this.name;
  getName = () => this.name;
  getMap = jest.fn().mockReturnValue(map);
  onClick = jest.fn();
  removeEntity = jest.fn();
  render = jest.fn();
  renderToTexture = jest
    .fn()
    .mockReturnValue(
      new Texture("testTexture", new Vector(), new Size(10, 10))
    );
  shouldShowBorder = jest.fn();
  startBackgroundMusic = jest.fn();
  update = jest.fn();
  unload = jest.fn();
}
