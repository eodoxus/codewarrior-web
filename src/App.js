import React, { Component } from "react";
import cx from "classnames";
import Indicators from "./components/indicators";
import Layout from "./components/layout";
import styles from "./App.scss";
import { AppModel } from "./data";
import Game from "./game";
import Url from "./lib/Url";

const DEFAULT_ROUTE = "home";
const GAME_WIDTH = 640;
const GAME_HEIGHT = 480;
const CHROME_HEIGHT = 76 + 40; // header + footer

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      name: "...",
      route: DEFAULT_ROUTE
    };
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: true,
      error: error
    });
  }

  componentDidMount() {
    window.addEventListener("resize", this.updateDimensions);
  }

  async componentWillMount() {
    let model = await new AppModel().load();
    this.setState(model.toPojo());
    if (!this.state.hasError) {
      this.setState({ isLoading: false });
    }
    this.updateDimensions();
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.updateDimensions);
  }

  onBorderClick = e => {
    e.currentTarget = this.sceneDirector.container;
    this.sceneDirector && this.sceneDirector.onClick(e);
  };

  render() {
    const content = this.state.hasError
      ? this.renderError()
      : this.renderGame();
    return (
      <div
        className={cx(
          this.state.centerHorizontally ? styles.centerVertically : "",
          this.state.centerVertically ? styles.centerVertically : "",
          styles.app
        )}
      >
        <Layout.Header
          email={this.state.email}
          name={this.state.name}
          phone={this.state.phone}
          portrait={this.state.avatar}
          slogan={this.state.slogan}
          url={this.state.home}
          hide={this.state.hideChrome}
        />
        <div className={styles.content}>{content}</div>
        <Layout.Footer copy={this.state.footer} hide={this.state.hideChrome} />
      </div>
    );
  }

  renderError() {
    return <div className={styles.error}>{this.state.error.message}</div>;
  }

  renderGame() {
    if (this.state.isLoading) {
      return <Indicators.Loader />;
    }
    let border;
    if (this.state.hasBorder) {
      border = (
        <div
          className={styles.border}
          style={{
            backgroundImage: `url(${Url.PUBLIC}/images/game-border.png)`
          }}
          onClick={this.onBorderClick}
        />
      );
    }
    return (
      <div className={styles.game}>
        {border}
        <Game.SceneDirector
          scene={this.state.route}
          width={GAME_WIDTH}
          height={GAME_HEIGHT}
          scale="2.5"
          ref={sceneDirector => (this.sceneDirector = sceneDirector)}
        />
      </div>
    );
  }

  updateDimensions = () => {
    const stateChanges = {
      hasBorder: false,
      centerHorizontally: false,
      centerVertically: false,
      hideChrome: false
    };
    if (window.innerWidth > GAME_WIDTH) {
      stateChanges.hasBorder = true;
      stateChanges.centerHorizontally = true;
    }
    if (window.innerHeight > GAME_HEIGHT) {
      stateChanges.centerVertically = true;
    }
    if (window.innerHeight <= GAME_HEIGHT + CHROME_HEIGHT) {
      stateChanges.hideChrome = true;
    }
    if (Object.keys(stateChanges).length) {
      this.setState(stateChanges);
    }
  };
}
