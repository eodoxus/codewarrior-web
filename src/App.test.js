import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App from "./App";
import { AppModel } from "./data";
import Scene from "./game/engine/Scene";

jest.mock("./game/engine/Graphics");

Scene.prototype.init = () => Promise.resolve();
Scene.prototype.render = () => true;
Scene.prototype.update = () => true;

describe("<App />", () => {
  beforeEach(function() {
    let model = new AppModel({
      avatar: "poo",
      email: "poody",
      footer: "footy",
      home: "/",
      name: "test",
      phone: "2234",
      slogan: "herro"
    });

    fetch.mockResponse(model.toJson());
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
  });

  it("matches expected snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
