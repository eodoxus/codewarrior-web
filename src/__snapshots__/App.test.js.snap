// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches expected snapshot 1`] = `
<div
  className=""
>
  <header
    className="collapsible header"
  >
    <div
      className={undefined}
    >
      <a
        className={undefined}
        href={undefined}
      >
        <img
          alt="... portrait"
          src={undefined}
        />
      </a>
    </div>
    <div
      className={undefined}
    >
      <h1
        className={undefined}
      >
        <a
          href={undefined}
        >
          ...
        </a>
      </h1>
      <h2
        className={undefined}
      />
    </div>
    <p
      className={undefined}
    >
      <a
        className="btn-sm btn btn-default"
        href="https://facebook.com/eodoxus"
        rel="noopener noreferrer"
        target="_blank"
      >
        <i
          className="fa fa-facebook"
        />
      </a>
      <a
        className="btn-sm btn btn-default"
        href="https://www.linkedin.com/in/jasongordon"
        rel="noopener noreferrer"
        target="_blank"
      >
        <i
          className="fa fa-linkedin"
        />
      </a>
      <a
        className="btn-sm btn btn-default"
        href="https://github.com/eodoxus"
        rel="noopener noreferrer"
        target="_blank"
      >
        <i
          className="fa fa-github"
        />
      </a>
      <span
        className={undefined}
      >
        <a
          className="btn-sm btn btn-default"
          href="tel:undefined"
          rel="noopener noreferrer"
          target="_top"
        >
          <i
            className="fa fa-phone"
          />
        </a>
      </span>
      <a
        className="btn-sm btn btn-default"
        href="mailto:undefined"
        rel="noopener noreferrer"
        target="_top"
      >
        <i
          className="fa fa-envelope"
        />
      </a>
    </p>
    <p
      className={undefined}
    >
       
      <br />
    </p>
  </header>
  <div
    className={undefined}
  >
    <div
      className=""
    >
      <svg
        height="40px"
        id="loader-1"
        style={
          Object {
            "enableBackground": "new 0 0 50 50",
          }
        }
        version="1.1"
        viewBox="0 0 50 50"
        width="40px"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <path
          d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
          fill="#000"
          transform="rotate(330 25 25)"
        >
          <animateTransform
            attributeName="transform"
            attributeType="xml"
            dur="0.6s"
            from="0 25 25"
            repeatCount="indefinite"
            to="360 25 25"
            type="rotate"
          />
        </path>
      </svg>
    </div>
  </div>
  <footer
    className="collapsible footer"
  >
    
  </footer>
</div>
`;
